### PR TITLE
[REF] account,*: simplify credit note wizard

### PR DESCRIPTION
**Pull request by sohailJ**
Description of the issue/feature this PR addresses:
simplification of the credit note wizard

Current behavior before PR:
First users select reverse option (3 radio buttons):
1) refund
2) cancel
3) modify

The reverse action is triggered when the users clicks on the "reverse" button

Desired behavior after PR is merged:
radio button are removed. There is now two buttons that trigger directly the reverse action with the desired option (refund or modify, cancel is removed)

task id: 3244377

[mockup](https://app.excalidraw.com/l/65VNwvy7c4X/2812i7OVfBl)
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by epictete — CHANGES_REQUESTED** [review]

File: addons/account/wizard/account_move_reversal.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ class AccountMoveReversal(models.TransientModel):
             ('entry', 'Journal Entry Date')
     ], required=True, default='custom')
     date = fields.Date(string='Reversal date', default=fields.Date.context_today)
-    reason = fields.Char(string='Reason')
+    reason = fields.Char(string='Reason displayed on Credit Notes')
     refund_method = fields.Selection(selection=[
             ('refund', 'Partial Refund'),
             ('cancel', 'Full Refund'),
```
[the remark this answers is not in the document]
Comment (sohailJ) [reply addons/account/wizard/account_move_reversal.py:25]: yes it is used in tests. I removed it and tests went red

File: addons/account/wizard/account_move_reversal_view.xml [hunk 2]
```diff
@@ -12,35 +12,13 @@
...
-                            <field name="date_mode" string="Reversal Date" widget="radio"/>
-                        </group>
-                        <group>
-                        <field name="journal_id" domain="[('id', 'in', available_journal_ids)]"/>
-                        <field name="date" string="Refund Date" attrs="{'invisible': ['|', ('move_type', 'not in', ('out_invoice', 'in_invoice')), ('date_mode', '!=', 'custom')], 'required':[('date_mode', '=', 'custom')]}"/>
-                        <field name="date" attrs="{'invisible': ['|', ('move_type', 'in', ('out_invoice', 'in_invoice')), ('date_mode', '!=', 'custom')], 'required':[('date_mode', '=', 'custom')]}"/>
-                        </group>
+                        <field name="reason" attrs="{'invisible': [('move_type', '=', 'entry')]}"/>
+                        <field name="journal_id" optional="True" domain="[('type', 'in', ['purchase', 'sale']), ('id', 'in', available_journal_ids)]" options="{'no_create': True, 'no_open': True}" context="{'default_type': 'purchase_refund'}" default="move_ids.journal_id and move_ids.journal_id[0] or False"/>
```
[the remark this answers is not in the document]
Comment (sohailJ) [reply addons/account/wizard/account_move_reversal_view.xml:16]: I don't know why I did this. It seems stupid to me now since I just had to keep it as is.

File: doc/cla/individual/sohailJ.md
[the remark this answers is not in the document]
Comment (qdp-odoo) [reply doc/cla/individual/sohailJ.md:1]: the email address linked to his github account isn't his odoo email address

File: addons/account/wizard/account_move_reversal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/account/wizard/account_move_reversal.py:25]: :see_no_evil: no, the option should be removed and the tests fixed (if the test is doing something else, you'd need to post the move, if the test is testing option 2 it must be deleted)

File: addons/account/wizard/account_move_reversal.py [hunk 3]
```diff
@@ -112,6 +112,7 @@ def _prepare_default_reversal(self, move):
     def reverse_moves(self):
         self.ensure_one()
         moves = self.move_ids
+        self.refund_method = self._context.get('default_refund_method') or self.refund_method
```
Comment (qdp-odoo) [inline addons/account/wizard/account_move_reversal.py:115]: do you need to check the context? the point of passing the key default_refund_method is that it will give proper default value at the record creation :point_up:

File: doc/cla/individual/sohailJ.md
[the remark this answers is not in the document]
Comment (epictete) [reply doc/cla/individual/sohailJ.md:1]: Indeed, we changed it in the account settings on GitHub and in the local git config, but `legal/cla` is still not happy :sweat_smile: 
Maybe we also need to change the commit author :man_shrugging:

File: addons/account/wizard/account_move_reversal.py (hunk 3 above)
Comment (qdp-odoo) [reply addons/account/wizard/account_move_reversal.py]: I just thought: maybe it's because of the override of default_get that my previous suggestion doesn't work

File: addons/account/tests/test_account_move_date_algorithm.py [hunk 4]
```diff
@@ -49,7 +49,7 @@ def _reverse_invoice(self, invoice):
             .create({
                 'journal_id': invoice.journal_id.id,
                 'reason': "no reason",
-                'refund_method': 'cancel',
+                'refund_method': 'refund',
```
Comment (epictete) [inline addons/account/tests/test_account_move_date_algorithm.py:52]: Shouldn't it be 'modify'?

File: addons/account/tests/test_account_move_in_invoice.py [hunk 5]
```diff
@@ -1178,43 +1178,42 @@ def test_in_invoice_create_refund(self):
         self.assertInvoiceValues(reverse_move, [
```
Comment (epictete) [inline addons/account/tests/test_account_move_in_invoice.py:1178]:
```suggestion
        self.assertInvoiceValues(new_invoice, [
```
The old variable name doesn't make sense anymore and creates confusion (several occurrences).

**Review by epictete — CHANGES_REQUESTED** [review]

File: addons/account/tests/test_account_move_in_invoice.py [hunk 6]
```diff
@@ -1178,43 +1178,42 @@ def test_in_invoice_create_refund(self):
...
-                'debit': 1128.0,
-                'credit': 0.0,
+                'amount_currency': -1128.0,
+                'debit': 0.0,
+                'credit': 1128.0,
                 'date_maturity': move_reversal.date,
             },
         ], {
             **self.move_vals,
-            'invoice_payment_term_id': None,
```
[the remark this answers is not in the document]
Comment (sohailJ) [reply addons/account/tests/test_account_move_in_invoice.py:1213]: because with `refund_method` set as `modify`, `invoice_payment_term_id` is not `None` and has a value

File: addons/account/tests/test_account_move_date_algorithm.py (hunk 4 above)
Comment (sohailJ) [reply addons/account/tests/test_account_move_date_algorithm.py:52]: the test just checks the date, 'modify' would change the behaviour

File: addons/l10n_latam_invoice_document/wizards/account_move_reversal_view.xml [hunk 7]
```diff
@@ -6,6 +6,7 @@
         <field name="model">account.move.reversal</field>
         <field name="inherit_id" ref="account.view_account_move_reversal"/>
         <field name="arch" type="xml">
+            <field name="refund_method" invisible="1"/>
```
[the remark this answers is not in the document]
Comment (sohailJ) [reply addons/l10n_latam_invoice_document/wizards/account_move_reversal_view.xml:9]: ok thanks, removing it

File: addons/sale/tests/test_sale_refund.py
[the remark this answers is not in the document]
Comment (sohailJ) [reply addons/sale/tests/test_sale_refund.py:1]: because there is also a test for modify here:
https://github.com/odoo/odoo/blob/b71144d1b48138c7d5e57c9c2d3912b52d1883f9/addons/sale/tests/test_sale_refund.py#L156

File: addons/account/tests/test_account_move_in_invoice.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (epictete) [reply addons/account/tests/test_account_move_in_invoice.py:1213]: Is that a problem?

[the remark this answers is not in the document]
Comment (sohailJ) [reply addons/account/tests/test_account_move_in_invoice.py:1213]: I can check that it has a value but I think that for the test the point was to check that the value is at None. Also I can not predict (if I am not wrong) the value here (id generated).

[the remark this answers is not in the document]
Comment (epictete) [reply addons/account/tests/test_account_move_in_invoice.py:1213]: The `id` of the payment term is deterministic, so you can use it safely :ok_hand: 
https://github.com/odoo/odoo/blob/10497b96c067c234553bf44d735e51414087bca9/addons/account/tests/common.py#L123

File: addons/account/tests/test_account_move_in_invoice.py [hunk 8]
```diff
@@ -1299,49 +1298,48 @@ def test_in_invoice_create_refund_multi_currency(self):
         self.assertInvoiceValues(reverse_move, [
```
Comment (epictete) [inline addons/account/tests/test_account_move_in_invoice.py:1298]:
```suggestion
        self.assertInvoiceValues(new_move, [
```

File: addons/account/tests/test_account_move_in_invoice.py [hunk 9]
```diff
@@ -1299,49 +1298,48 @@ def test_in_invoice_create_refund_multi_currency(self):
...
                 'currency_id': self.currency_data['currency'].id,
-                'debit': 564.0,
-                'credit': 0.0,
+                'debit': 0.0,
+                'credit': 564.0,
                 'date_maturity': move_reversal.date,
             },
         ], {
             **self.move_vals,
-            'invoice_payment_term_id': None,
```
Comment (epictete) [inline addons/account/tests/test_account_move_in_invoice.py:1339]: Restore field.

File: addons/account/tests/test_account_move_out_invoice.py [hunk 10]
```diff
@@ -1795,43 +1795,42 @@ def test_out_invoice_create_refund(self):
         self.assertInvoiceValues(reverse_move, [
```
Comment (epictete) [inline addons/account/tests/test_account_move_out_invoice.py:1795]:
```suggestion
        self.assertInvoiceValues(new_move, [
```

File: addons/account/tests/test_account_move_out_invoice.py [hunk 11]
```diff
@@ -1795,43 +1795,42 @@ def test_out_invoice_create_refund(self):
...
-                'debit': 0.0,
-                'credit': 1410.0,
+                'amount_currency': 1410.0,
+                'debit': 1410.0,
+                'credit': 0.0,
                 'date_maturity': move_reversal.date,
             },
         ], {
             **self.move_vals,
-            'invoice_payment_term_id': None,
```
Comment (epictete) [inline addons/account/tests/test_account_move_out_invoice.py:1830]: Restore field.

File: addons/account/tests/test_account_move_out_invoice.py [hunk 12]
```diff
@@ -1916,49 +1915,48 @@ def test_out_invoice_create_refund_multi_currency(self):
         self.assertInvoiceValues(reverse_move, [
```
Comment (epictete) [inline addons/account/tests/test_account_move_out_invoice.py:1915]:
```suggestion
        self.assertInvoiceValues(new_move, [
```

File: addons/account/tests/test_account_move_out_invoice.py [hunk 13]
```diff
@@ -1916,49 +1915,48 @@ def test_out_invoice_create_refund_multi_currency(self):
...
                 'currency_id': self.currency_data['currency'].id,
-                'debit': 0.0,
-                'credit': 705.0,
+                'debit': 705.0,
+                'credit': 0.0,
                 'date_maturity': move_reversal.date,
             },
         ], {
             **self.move_vals,
-            'invoice_payment_term_id': None,
```
Comment (epictete) [inline addons/account/tests/test_account_move_out_invoice.py:1956]: Restore field.

File: addons/account/wizard/account_move_reversal_view.xml [hunk 14]
```diff
@@ -11,36 +11,15 @@
...
-                            <field name="reason" attrs="{'invisible': [('move_type', '=', 'entry')]}"/>
-                            <field name="date_mode" string="Reversal Date" widget="radio"/>
-                        </group>
-                        <group>
+                        <field name="reason" attrs="{'invisible': [('move_type', '=', 'entry')]}"/>
                         <field name="journal_id" domain="[('id', 'in', available_journal_ids)]"/>
-                        <field name="date" string="Refund Date" attrs="{'invisible': ['|', ('move_type', 'not in', ('out_invoice', 'in_invoice')), ('date_mode', '!=', 'custom')], 'required':[('date_mode', '=', 'custom')]}"/>
-                        <field name="date" attrs="{'invisible': ['|', ('move_type', 'in', ('out_invoice', 'in_invoice')), ('date_mode', '!=', 'custom')], 'required':[('date_mode', '=', 'custom')]}"/>
-                        </group>
+                        <field name="date" widget="date"/>
```
Comment (epictete) [inline addons/account/wizard/account_move_reversal_view.xml:18]: Should be required.

File: addons/purchase_stock/tests/test_anglo_saxon_valuation_reconciliation.py [hunk 15]
```diff
@@ -106,15 +106,15 @@ def test_invoice_shipment(self):
...
+            'refund_method': 'modify',
             'date': '2018-03-15',
             'journal_id': invoice.journal_id.id,
         })
-        refund_invoice = self.env['account.move'].browse(refund_invoice_wiz.reverse_moves()['res_id'])
-
+        new_invoice = self.env['account.move'].browse(refund_invoice_wiz.reverse_moves()['res_id'])
+        refund_invoice = invoice.reversal_move_id
         # Check the result
-        self.assertEqual(invoice.payment_state, 'reversed', "Invoice should be in 'reversed' state")
```
Comment (epictete) [inline addons/purchase_stock/tests/test_anglo_saxon_valuation_reconciliation.py:116]: Why did you remove it?

Comment (sohailJ) [reply addons/purchase_stock/tests/test_anglo_saxon_valuation_reconciliation.py:116]: restored

**Review by epictete — APPROVED** [review]

**Review by epictete — APPROVED** [review]
@robodoo r+

File: addons/account/tests/test_account_move_out_invoice.py [hunk 16]
```diff
@@ -1906,59 +1906,59 @@ def test_out_invoice_create_refund_multi_currency(self):
         move_reversal = self.env['account.move.reversal'].with_context(active_model="account.move", active_ids=self.invoice.ids).create({
             'date': fields.Date.from_string('2017-01-01'),
             'reason': 'no reason',
-            'refund_method': 'cancel',
+            'refund_method': 'modify',
```
Comment (qdp-odoo) [inline addons/account/tests/test_account_move_out_invoice.py:1909]: refuuund :facepalm:

File: addons/account/tests/test_account_move_in_invoice.py [hunk 17]
```diff
@@ -1168,53 +1168,53 @@ def test_in_invoice_create_refund(self):
         move_reversal = self.env['account.move.reversal'].with_context(active_model="account.move", active_ids=self.invoice.ids).create({
             'date': fields.Date.from_string('2019-02-01'),
             'reason': 'no reason again',
-            'refund_method': 'cancel',
+            'refund_method': 'modify',
```
Comment (qdp-odoo) [inline addons/account/tests/test_account_move_in_invoice.py:1171]: cancel => refund + post

:facepalm:

File: addons/sale_stock/tests/test_sale_stock_report.py [hunk 18]
```diff
@@ -360,68 +360,6 @@ def test_invoice_with_several_returns(self):
         self.assertRegex(text, r'Product By Lot\n2.00Units\nLOT0003', "There should be a line that specifies 2 x LOT0003")
         self.assertNotIn('LOT0001', text)
 
-    def test_refund_cancel_invoices(self):
```
Comment (qdp-odoo) [inline addons/sale_stock/tests/test_sale_stock_report.py:363]: you can't remove this test??

File: addons/account/models/account_move.py [hunk 19]
```diff
@@ -3498,11 +3498,22 @@ def _post(self, soft=True):
             if wrong_lines:
                 wrong_lines.write({'partner_id': invoice.commercial_partner_id.id})
 
+        # reconcile if state is in draft and move has reversal_entry_id set
+        draft_refunds = to_post.filtered(lambda refund: refund.reversed_entry_id and refund.state == 'draft')
```
Comment (epictete) [inline addons/account/models/account_move.py:3502]:
```suggestion
        draft_reverse_moves = to_post.filtered(lambda move: move.reversed_entry_id)
```
It could also be journal entries (so not a refund).
And you don't need to check the state, it cannot be posted.
Though it could technically be cancel, we should also change the check done above (line 3459) to include cancelled moves.

File: addons/account/models/account_move.py [hunk 20]
```diff
@@ -3498,11 +3498,22 @@ def _post(self, soft=True):
...
             'posted_before': True,
         })
 
+        for move, reverse_move in zip(draft_refunds.reversed_entry_id, draft_refunds):
+            group = (move.line_ids + reverse_move.line_ids) \
+                .filtered(lambda l: not l.reconciled) \
+                .grouped(lambda l: (l.account_id, l.currency_id))
+            for (account, _currency), lines in group.items():
+                if account.reconcile or account.account_type in ('asset_cash', 'liability_credit_card'):
+                    lines.with_context(move_reverse_cancel=True).reconcile()
```
Comment (epictete) [inline addons/account/models/account_move.py:3515]:
```suggestion
        for move, reverse_move in zip(draft_refunds.reversed_entry_id, draft_refunds):
            group = (move.line_ids + reverse_move.line_ids) \
                .filtered(lambda l: not l.reconciled) \
                .grouped(lambda l: (l.account_id, l.currency_id))
            for (account, _currency), lines in group.items():
                if account.reconcile or account.account_type in ('asset_cash', 'liability_credit_card'):
                    lines.reconcile()
```
This code is very similar (if not identical) to what is done in `_reverse_moves()`.
So instead of duplicating it, you might as well extract that logic in a dedicated method that you can call in both instances.
And forcing `move_reverse_cancel` to ´True` should not be necessary but could be problematic.

File: addons/account/wizard/account_move_reversal.py [hunk 21]
```diff
@@ -14,21 +14,16 @@ class AccountMoveReversal(models.TransientModel):
...
-    ], required=True, default='custom')
     date = fields.Date(string='Reversal date', default=fields.Date.context_today)
-    reason = fields.Char(string='Reason')
+    reason = fields.Char(string='Reason displayed on Credit Note')
     refund_method = fields.Selection(selection=[
             ('refund', 'Partial Refund'),
-            ('cancel', 'Full Refund'),
             ('modify', 'Full refund and new draft invoice')
         ], string='Credit Method', required=True,
         help='Choose how you want to credit this invoice. You cannot "modify" nor "cancel" if the invoice is already reconciled.')
```
Comment (epictete) [inline addons/account/wizard/account_move_reversal.py:23]: This field is no longer used in the wizard, so we should remove it :broom: :slightly_smiling_face:

**Review by epictete — CHANGES_REQUESTED** [review]

**Comment by qdp-odoo** [comment]
@robodoo r+ rebase-ff

File: addons/account/models/account_move.py [hunk 22]
```diff
@@ -3342,6 +3342,20 @@ def _get_reconciled_invoices_partials(self):
                 exchange_diff_moves.append(partial.exchange_move_id.id)
         return invoice_partials, exchange_diff_moves
 
+    def _reconcile_reversed_moves(self, moves, reverse_moves, move_reverse_cancel):
```
Comment (epictete) [inline addons/account/models/account_move.py:3345]: Why are you using `moves` as parameter, when you could just call the method on `moves` instead of `self`?

File: addons/account/models/account_move.py [hunk 23]
```diff
@@ -3342,6 +3342,20 @@ def _get_reconciled_invoices_partials(self):
...
+        ''' Reconciles moves and reverse moves
```
Comment (epictete) [inline addons/account/models/account_move.py:3346]: Your doc string is identical to the method name, it could be a bit more detailed :wink:

**Review by epictete — COMMENTED** [review]
Is your second commit a `[REF]` or an `[IMP]`?

**Comment by epictete** [comment]
@robodoo r+
